packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,222 @@
+import { styled, useThemeProps } from '@material-ui/core/styles';
+import { getMasonryItemUtilityClass } from './masonryItemClasses';
+import MasonryContext from '../Masonry/MasonryContext';
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryItemUtilityClass, classes);
+};
+
+export const style = ({ styleProps, theme }) => {
+  let styles = {
+    width: '100%',
+    '& > *': {
+      // all contents should have a width of 100%
+      width: '100%',
+      boxSizing: 'inherit',
+      ...(styleProps.isSSR && { height: '100%' }),
+      ...(styleProps.isImage && !styleProps.isSSR && { height: styleProps.height }),
+    },
+    visibility: styleProps.height ? 'visible' : 'hidden',
+    gridColumnEnd: `span ${styleProps.columnSpan}`,
+    boxSizing: 'inherit',
+  };
+
+  const base = {};
+  Object.keys(theme.breakpoints.values).forEach((breakpoint) => {
+    if (styleProps.spacing[breakpoint] != null) {
+      base[breakpoint] = true;
+    }
+  });
+  const spacingValues = resolveBreakpointValues({ values: styleProps.spacing, base });
+  const transformer = createUnarySpacing(theme);
+  const styleFromPropValue = (propValue) => {
+    const gap = styleProps.height ? Number(getValue(transformer, propValue).replace('px', '')) : 0;
+    const rowSpan = styleProps.height ? Math.ceil(styleProps.height + gap) : 0;
+    return {
+      gridRowEnd: `span ${rowSpan}`,
+      paddingBottom: gap - 1,
+    };
+  };
+  styles = { ...styles, ...handleBreakpoints({ theme }, spacingValues, styleFromPropValue) };
+
+  return styles;
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(style);
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+
+  const { spacing = 1 } = React.useContext(MasonryContext);
+  const isImage = (element) => {
+    return element.type === 'img';
+  };
+  const {
+    children,
+    className,
+    component = 'div',
+    columnSpan = 1,
+    height = isImage(children) ? 1 : undefined,
+    // images should have defined dimensions for browser to correctly lay out and paint.
+    // browser can't handle a large number of images without pre-defined dimensions
+    ...other
+  } = props;
+  const isSSR = !((isImage(children) && height === 1) || height === undefined);
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    isSSR,
+    isImage: isImage(children),
+    spacing,
+    columnSpan,
+    height: height < 0 ? 0 : height, // MasonryItems to which negative or zero height is passed will be hidden
+  });
+
+  const classes = useUtilityClasses(styleProps);
+  const resizeObserver = React.useRef(null);
```

Comment:
Looks like the ref is not used. We never actually read from `resizeObserver.current`

Reply:
We use inside `useEffect()`. https://github.com/mui-org/material-ui/pull/27439/commits/92252de23472aa6063fe2e65793b08a538299d4a#diff-f3824fe7ebb878266a976503904c0d5681fe98068e49820cf469d979ed572e18R144

Reply:
But, I think we can define it inside useEffect and not use ref.